tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c

```diff
@@ -0,0 +1,650 @@
+	TEST_ASSERT(buf->len == hdr->len, "buflen = %d != hdrlen %d",
+	       buf->len, hdr->len);
+
+	TEST_ASSERT(cid == BT_L2CAP_CID_ATT, "We only support (U)ATT");
+
+	/* (U)ATT PDU */
+	handle_att(buf);
+}
+
+static void handle_acl(struct net_buf *buf)
+{
+	struct bt_hci_acl_hdr *hdr;
+	uint16_t len, handle;
+	uint8_t flags;
+
+	hdr = net_buf_pull_mem(buf, sizeof(*hdr));
+	len = sys_le16_to_cpu(hdr->len);
+	handle = sys_le16_to_cpu(hdr->handle);
+
+	flags = bt_acl_flags(handle);
+	handle = bt_acl_handle(handle);
+
+	TEST_ASSERT(flags == BT_ACL_START,
+	       "Fragmentation not supported");
+
+	LOG_DBG("ACL: conn %d len %d flags %d", handle, len, flags);
+	LOG_HEXDUMP_DBG(buf->data, buf->len, "HCI ACL");
+
+	handle_l2cap(buf);
+}
+
+static void recv(struct net_buf *buf)
+{
+	LOG_HEXDUMP_DBG(buf->data, buf->len, "HCI RX");
+
+	uint8_t code = buf->data[0];
+
+	if (bt_buf_get_type(buf) == BT_BUF_EVT) {
+		switch (code) {
+		case BT_HCI_EVT_CMD_COMPLETE:
+		case BT_HCI_EVT_CMD_STATUS:
+			handle_cmd_complete(buf);
+			break;
+		case BT_HCI_EVT_LE_META_EVENT:
+			handle_meta_event(buf);
+			break;
+		case BT_HCI_EVT_DISCONN_COMPLETE:
+			UNSET_FLAG(is_connected);
+			break;
+		case BT_HCI_EVT_NUM_COMPLETED_PACKETS:
+			handle_ncp(buf);
+			break;
+		default:
+			LOG_ERR("unhandled msg %x", code);
+			LOG_HEXDUMP_ERR(buf->data, buf->len, "HCI EVT");
+		}
+
+		/* handlers should take a ref if they want to access the buffer
+		 * later
+		 */
+		net_buf_unref(buf);
+		return;
+	}
+
+	if (bt_buf_get_type(buf) == BT_BUF_ACL_IN) {
+		handle_acl(buf);
+		net_buf_unref(buf);
+		return;
+	}
+
+	LOG_ERR("HCI RX (not data or event)");
+	net_buf_unref(buf);
+}
+
+static void send_cmd(uint16_t opcode, struct net_buf *cmd, struct net_buf **rsp)
+{
+	LOG_DBG("opcode %x", opcode);
+
+	if (!cmd) {
+		cmd = bt_hci_cmd_create(opcode, 0);
+	}
+
+	k_sem_take(&cmd_sem, K_FOREVER);
+	TEST_ASSERT(active_opcode == 0xFFFF, "");
+
+	active_opcode = opcode;
+
+	LOG_HEXDUMP_DBG(cmd->data, cmd->len, "HCI TX");
+	bt_send(cmd);
+
+	/* Wait until the command completes */
+	k_sem_take(&cmd_sem, K_FOREVER);
+	k_sem_give(&cmd_sem);
+
+	net_buf_unref(cmd);
+
+	/* return response. it's okay if cmd_rsp gets overwritten, since the app
+	 * gets the ref to the underlying buffer when this fn returns.
+	 */
+	if (rsp) {
+		*rsp = cmd_rsp;
+	} else {
+		net_buf_unref(cmd_rsp);
+		cmd_rsp = NULL;
+	}
+}
+
+static K_THREAD_STACK_DEFINE(rx_thread_stack, 1024);
+static struct k_thread rx_thread_data;
+
+static void rx_thread(void *p1, void *p2, void *p3)
+{
+	LOG_DBG("start HCI rx");
+
+	while (true) {
+		struct net_buf *buf;
+
+		/* Wait until a buffer is available */
+		buf = net_buf_get(&rx_queue, K_FOREVER);
+		recv(buf);
+	}
+}
+
+static void le_read_buffer_size_complete(struct net_buf *rsp)
+{
+	struct bt_hci_rp_le_read_buffer_size *rp = (void *)rsp->data;
+
+	LOG_DBG("status 0x%02x", rp->status);
+	LOG_DBG("max len %d max num %d", rp->le_max_len, rp->le_max_num);
```

Comment:
We should assert that `rp->le_max_len` is large enough to send a long l2cap packet. The other option is to rely on that the controller supports this length and validates when it receives it. If we do that we do not have to issue `BT_HCI_OP_LE_READ_BUFFER_SIZE`

Reply:
what do you mean by long l2cap packet? isn't it decoupled from controller buffer size?

Reply:
I meant that the buffer size is large enough to fit an L2CAP packet in a single HCI Data packet right